### ISSUE TITLE
WIP, DOC: draft PR description for Emma's sklearn PR

### ISSUE THREAD

**Issue by tylerjereddy** [lanl/GFDL#13]
Related to gh-1 and the `scikit-learn` estimator PR.

Some notes up front:

- [ ] this is a work in progress and I encourage the team to improve on it--I believe the plan is for Emma to open the sklearn PR, so this is really just intended to help make that process smoother
- [x] I suggest that Emma opens a draft/WIP PR against `scikit-learn` `main` branch on her fork of `scikit-learn` (NOT their actual repo to start with), so that we do not cause noise in their primary repo while we iterate on the code and presentation/description, make sure the CI is passing, etc.
- [ ] typically I'd suggest having a separate commit that re-enables `scikit-learn` CI testing on forks, so that we can see what testing issues we'll have against their library, but also be ready to revert those extraneous CI config changes when it comes time to open the PR against their repo proper
- [ ] it should be fine to use whatever approach you want to get the relevant estimators over into an `sklearn` branch (i.e., pasting them into whatever appropriate modules based on some `git grep` analysis of how MLP is organized over there)
- [ ] I would ask that a reasonable effort is made to give Navamita and I a little bit of git history credit--if you condense everything to a single new commit, you can just include us as co-authors on the commit as described at: https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors
- [ ] Kostas said he might be able to help with this description write up as well--feel free to reach out to him, try to do that early on
- [ ] If we want to reach out to a sklearn developer, I've worked with Virgil Chan (https://scikit-learn.org/stable/about.html#contributor-experience-team) in the past; Emma could ping him for initial feedback when the PR against their repo is opened
- [ ] check their contributing guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
- [ ] follow their pull request checklist: https://scikit-learn.org/dev/developers/contributing.html#pull-request-checklist

And now some draft prose proper for the PR description is below. This is only a suggestion--feel free to modify it quite radically, especially if you see better examples of big changes/additions in the sklearn repo. It is not complete, and is only meant to get things rolling.

-----------------------

## Introduction

As part of a new 3-year DOE SC ASCR applied math grant, we've been performing research on gradient free neural network architectures that strike a balance between affordable/efficient (i.e., non-GPU) hardware/compute cost and model accuracy. We found that the open source scientific Python community was lacking some very fundamental estimators in this regard. In particular, there are currently no well-maintained multi-layer random vector functional link (RVFL) network  or extreme learning machine (ELM) libraries that we are aware of. https://github.com/thieu1995/GrafoRVFL currently use a GPL license, and only supports single-layer RVFL networks. Here, we propose to add support for two types of fundamental gradient-free estimators: `RVFLClassifier` and `RVFLRegressor`, which also offer parameters to switch over to ELM. A wider range of estimators is being developed at https://github.com/lanl/GFDL, but the more specialized variants may not be of sufficiently broad interest for this project at this time. Our team consists of myself (Emma), Navamita Ray, and Tyler Reddy (CS team at LANL); and Kyle Luh (CU Boulder) and Kostas (Boston University) on the mathematics side.

## Relevance and "fit" of the new estimators

A key question is whether these two estimators are of sufficiently broad interest for inclusion in `scikit-learn`. Their gradient-free nature means that they are generally quite efficient and don't require specialized hardware, consistent with a large number of heavily-used estimators like `RandomForestClassifier`. In terms of the breadth of their usage in the community, the table below summarizes a small sampling of relevant papers and citation counts (there are many others...).

Topic| DOI | Citations (Google Scholar, Jan 18, 2026)
-- | -- | --
Evaluating RVFLs on common datasets | https://doi.org/10.1016/j.ins.2015.09.025 | 498
Application of RVFLs to electricity load forecasting | https://doi.org/10.1016/j.ins.2015.11.039 | 336
Exploration of deep RVFL architectures | https://doi.org/10.1016/j.patcog.2021.107978 | 277
Introduction of ELMs (the basic gradient-free NN concept) | https://doi.org/10.1016/j.neucom.2005.12.126 | 16,475

## How do these new estimators work and avoid gradients/backpropagation?

The tradeoff for avoiding backpropagation/GPUs is largely related to randomness--the weights of all neurons are randomly set and frozen during training, with the exception of the connection between the final layer and the output. This simplification allows a closed form solution that can proceed either via Moore-Poore pseudoinverse (i.e., `np.pinv`) or via the various algorithms underlying ridge regression (literally `Ridge` in `scikit-learn`). A simple diagram of an RVFL demonstrates one additional twist--a direct link between the inputs and the outputs (source: https://doi.org/10.1016/j.patcog.2021.107978) that the community suggests is a form of regularization:

<img width="386" height="241" alt="Image" src="https://github.com/user-attachments/assets/0c272012-95c7-4b4a-9665-9f2feccd2045" />

As a result of the randomness in the training of these networks, we have to provide a wide range of options for setting the initial weights of the hidden layers, since they are frozen once set and can substantially affect estimator performance.

## API and design considerations

One nice thing about the design is that it provides direct access to two of the most popular gradient-free neural networks (RVFL and ELM), with interchange between them available via a single parameter (when `direct_links=True` you have an RVFL, otherwise an ELM). Some debate could be had about whether it makes sense to only offer the classical forms (single-layer only) of ELM and RVFL, but we ultimately found considerable literature exploring the deep/multi-layer versions of each type of architecture, and wanted access to well-maintained code that could be used to reproduce such studies.

Some debate may be needed on the pass-through to `Ridge` arguments we use internally. `solver='auto'` is currently used, and may be a sensible default. However, as we have experienced in comparing with numerical experiments with mathematicians, it is likely of interest to researchers to have the ability to control the solver used by `Ridge` to be able to closely compare these gradient-free models with i.e., backpropagated models that use specific solvers.

## Performance of the estimators on common datasets

In the table below, we report some estimator performances for these new models on common classification and regression datasets. Note that these are not refined results (in the absence of substantial hyperopt).

Dataset |  Estimator  | Metric = Value  
-- | -- | --
Boston Housing | RVFLRegressor | $R^2 = $
X                         | RVFLClassifier | ROC AUC =

### Licensing Considerations

We did not use LLMs to draft or to assist in the drafting of this code, and did not consult copyleft/GPL-licensed code, apart from checking the executed output from the `graforvfl` library. Note that we've also received an informal positive response from the grafo team that they will switch to a more permissive license in the future to alleviate concerns about this: https://github.com/thieu1995/GrafoRVFL/issues/1#issuecomment-3253116472.

**Comment by eiviani-lanl** [comment]
Question: `scikit-learn` doesn't seem to currently have any models where the distribution of weights can be set explicitly during initialization (correct if wrong).  The path of least resistance would be to allow the user to optionally supply their own weights as is done in the `SGDClassifier` `fit`. We could also move the weight initializers back into the model structure. I'd hesitate adding a new module for setting weights as that would just be another change to justify in the PR.
